streamcopy/copy.go

```diff
@@ -0,0 +1,57 @@
+package streamcopy
+
+import (
+	"bufio"
+	"io"
+	"time"
+
+	"github.com/containers/image/types"
+)
+
+const megaByte = float64(1024 * 1024)
+const readerSize = 32768
+
+// WithProgress implements io.Copy with a buffered reader, then measures
+// reports to the offsetFunc with the count processed at the interval.
+// If io.EOF is not returned then the error is returned. Otherwise, it is nil.
+// WithProgress closes the writer after finishing.
+func WithProgress(writer io.WriteCloser, reader io.Reader, artifact types.BlobInfo, interval time.Duration, signal chan types.SignalProperties) error {
+	defer writer.Close()
```

Comment:
It would be more reliable to default to a `CloseWithError` and only to set `err` to `nil` if we _know_ we are successful; we can exit the function not only by `return`, but also by a `panic` anywhere in the called code.

See `compressGoroutine`; I don’t know that it is perfect but it does seem to be more paranoid :)